src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/repository/OffenderProgramProfileRepository.kt

```diff
@@ -20,4 +23,24 @@ interface OffenderProgramProfileRepository : JpaRepository<OffenderProgramProfil
   ): OffenderProgramProfile?
 
   fun findByCourseActivity(courseActivity: CourseActivity): List<OffenderProgramProfile>
+
+  @Query(
+    value = """
+       select opp.offenderProgramReferenceId
+       from OffenderProgramProfile opp
+       join OffenderBooking ob on opp.offenderBooking = ob
+       join CourseActivity ca on opp.courseActivity.courseActivityId = ca.courseActivityId 
+       where opp.prison.id = :prisonId 
+       and opp.programStatus.code = 'ALLOC'
+       and opp.startDate <= current_date
+       and (opp.endDate is null or opp.endDate > current_date)
+       and ob.active = true
+       and ob.location.id = :prisonId
+       and ca.prison.id = :prisonId
+       and ca.active = true
+       and ca.scheduleStartDate <= current_date
+       and (ca.scheduleEndDate is null or ca.scheduleEndDate > current_date)
+  """,
```

Comment:
One downside is that we now have 2 queries, one for migrating course activities and one for migrating course allocations. We need to be careful that these don't diverge. Maybe we should restructure the tests that prove they haven't diverged? 🤔

Reply:
If only you used a specification .... (that is a joke)

Reply:
Oh no you actually have a point! 😆